In [3]:

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline


###scikit Learn Modules needed for Logistic Regression
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split,GridSearchCV
from sklearn.preprocessing import LabelEncoder,MinMaxScaler,OneHotEncoder,StandardScaler
from sklearn.preprocessing import LabelEncoder,MinMaxScaler,OneHotEncoder,StandardScaler 
from sklearn.metrics import confusion_matrix
from sklearn.impute import SimpleImputer
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer

In [59]:
df=pd.read_csv('/content/HomeLoan.csv')
df

Age  Credit_Rating  Children        Loan
0   19.0         27.900        0.0  16884.924
1   18.0         42.130        1.0  1725.5523
2   28.0         33.000        3.0   4449.462
3   33.0         22.705        0.0   21984.47
4   32.0         28.880        0.0   3866.855
5   31.0         26.400        1.0  18430.356
6   24.0         27.410        0.0   9836.145
7   27.0         31.160        1.0  18110.345
8   30.0         29.110        1.0  12810.116
9    NaN            NaN        NaN        NaN
10   NaN            NaN        NaN        NaN
11  41.0         23.500        NaN          ?

In [86]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12 entries, 0 to 11
Data columns (total 4 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Age            10 non-null     float64
 1   Credit_Rating  10 non-null     float64
 2   Children       9 non-null      float64
 3   Loan           10 non-null     object 
dtypes: float64(3), object(1)
memory usage: 512.0+ bytes


In [60]:
type(df)

pandas.core.frame.DataFrame

In [61]:
# number of rows and columns
df.shape

(12, 4)

In [62]:
# statistical measures
df.describe()

Age  Credit_Rating  Children 
count  10.000000      10.000000   9.000000
mean   28.300000      29.219500   0.777778
std     6.832114       5.507392   0.971825
min    18.000000      22.705000   0.000000
25%    24.750000      26.652500   0.000000
50%    29.000000      28.390000   1.000000
75%    31.750000      30.647500   1.000000
max    41.000000      42.130000   3.000000

In [63]:
# number of missing values in each column
df.isnull().sum()

Age              2
Credit_Rating    2
Children         3
Loan             2
dtype: int64

In [64]:
# dropping the missing values
df1 = df.dropna()

In [66]:
# number of missing values in each column
df1.isnull().sum()

Age              0
Credit_Rating    0
Children         0
Loan             0
dtype: int64

In [48]:
#df = df.apply (pd.to_numeric, errors='coerce')
#df = df.dropna()
#df.isnull().any()

In [67]:
df1

Age  Credit_Rating  Children        Loan
0  19.0         27.900        0.0  16884.924
1  18.0         42.130        1.0  1725.5523
2  28.0         33.000        3.0   4449.462
3  33.0         22.705        0.0   21984.47
4  32.0         28.880        0.0   3866.855
5  31.0         26.400        1.0  18430.356
6  24.0         27.410        0.0   9836.145
7  27.0         31.160        1.0  18110.345
8  30.0         29.110        1.0  12810.116

In [71]:
X=df1.iloc[:,:-1]
Y=df1['Loan']

In [74]:
# Train & Test split
x_train, x_test, y_train, y_test = train_test_split( X,Y,test_size=0.20,random_state=21)

print('Shape of Training Xs:{}'.format(x_train.shape))
print('Shape of Test Xs:{}'.format(x_test.shape))
print('Shape of Training y:{}'.format(y_train.shape))
print('Shape of Test y:{}'.format(y_test.shape))

Shape of Training Xs:(7, 3)
Shape of Test Xs:(2, 3)
Shape of Training y:(7,)
Shape of Test y:(2,)


In [87]:
from sklearn.linear_model import LinearRegression
lr = LinearRegression()
lr.fit(x_train, y_train)
y_pred= lr.predict(x_test)
print("Accuracy on Traing set: ",lr.score(x_train,y_train))
print("Accuracy on Testing set: ",lr.score(x_test,y_test))

Accuracy on Traing set:  0.6857505187294235
Accuracy on Testing set:  -160.31268448059268


In [105]:
mean_absolute_percentage_error(y_test, y_pred)

4.934010358348127

In [88]:
df_prediction = pd.DataFrame({"Actual Selling Price":y_test,"Predicted_Price":y_pred})

In [77]:
df_prediction.head(n=10)

Actual Selling Price  Predicted_Price
1            1725.5523     -6346.505355
2             4449.462     27542.457942

In [90]:
X.columns

Index(['Age', 'Credit_Rating', 'Children '], dtype='object')

In [91]:
data_new={'Age':41, 'Credit_Rating':23.50,'Children':0 } 
index=[1]#serial number
my_data=pd.DataFrame(data_new,index)

In [92]:
print(my_data)

   Age  Credit_Rating  Children
1   41           23.5         0


In [94]:
my_new_loan=lr.predict(my_data)
rounded_loan = np.round(my_new_loan, 2)
print(f" The predicted price for the given data is :{rounded_loan}")


 The predicted price for the given data is :[11258.66]


/usr/local/lib/python3.8/dist-packages/sklearn/base.py:493: FutureWarning: The feature names should match those that were passed during fit. Starting version 1.2, an error will be raised.
Feature names unseen at fit time:
- Children
Feature names seen at fit time, yet now missing:
- Children 

  warnings.warn(message, FutureWarning)


**Random Forest Regression**

In [103]:
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_absolute_percentage_error
 
model_RFR = RandomForestRegressor(n_estimators=10)
model_RFR.fit(x_train, y_train)
y_pred_rf = model_RFR.predict(x_test)
 
mean_absolute_percentage_error(y_test, y_pred_rf)

5.102505389603073

**SVM – Support vector Machine**

SVM can be used for both regression and classification model. It finds the hyperplane in the n-dimensional plane.

In [104]:
from sklearn import svm
from sklearn.svm import SVC
from sklearn.metrics import mean_absolute_percentage_error
 
model_SVR = svm.SVR()
model_SVR.fit(x_train,y_train)
y_pred_sv = model_SVR.predict(x_test)
 
print(mean_absolute_percentage_error(y_test, y_pred_sv))

5.790028052788781


Conclusion 
Clearly, Linear Regression model is giving better accuracy as the mean absolute error is the least among all the other regressor models i.e. 4.9 approx.